lambdas/WavWrapLambdaFunction/index.js

```diff
@@ -0,0 +1,162 @@
+  ) {
+    delete event.Task.Destination.Parameters.Bucket;
+    delete event.Task.Destination.Parameters.Key;
+    delete event.Task.Destination.Parameters.Body;
+
+    Object.assign(params, event.Task.Destination.Parameters);
+  }
+
+  // Upload the resulting file to the destination in S3
+  const uploadStart = process.hrtime();
+  await s3writer.upload(params).promise();
+
+  const uploadEnd = process.hrtime(uploadStart);
+  console.log(
+    JSON.stringify({
+      msg: 'Finished S3 upload',
+      duration: `${uploadEnd[0]} s ${uploadEnd[1] / 1000000} ms`,
+    }),
+  );
+}
+
+exports.handler = async (event) => {
+  console.log(JSON.stringify({ msg: 'State input', input: event }));
+
+  const s3 = new AWS.S3({ apiVersion: '2006-03-01' });
+  const sts = new AWS.STS({ apiVersion: '2011-06-15' });
+
+  // Fetch the source file artifact from S3
+  console.log(
+    JSON.stringify({
+      msg: 'Fetching artifact from S3',
+      s3: `${event.Artifact.BucketName}/${event.Artifact.ObjectKey}`,
+    }),
+  );
+
+  const s3start = process.hrtime();
+  let s3Object = null;
+  try {
+    s3Object = await s3
+      .getObject({
+        Bucket: event.Artifact.BucketName,
+        Key: event.Artifact.ObjectKey,
+      })
+      .promise();
+  } catch (error) {
+    console.log(
+      'S3 getObject failed',
+      event.Artifact.BucketName,
+      event.Artifact.ObjectKey,
+      error,
+    );
+    throw error;
+  }
+
+  const s3end = process.hrtime(s3start);
+  console.log(
+    JSON.stringify({
+      msg: 'Fetched artifact from S3',
+      duration: `${s3end[0]} s ${s3end[1] / 1000000} ms`,
+    }),
+  );
+
+  // create the wav object
+  const wav = new wavefile.WaveFile();
+  wav.fromMpeg(s3Object.Body);
+
+  // If there are chunks passed in, iterate through each
+  const resultChunks = [];
+  event.Task.Chunks.forEach((taskChunk) => {
+    // Set data if this is a chunk supported by wavefile
+    const chunkId = taskChunk.ChunkId;
+    if (Object.prototype.hasOwnProperty.call(wav, chunkId)) {
+      Object.keys(taskChunk).forEach((taskKey) => {
+        const wavKey = camelize(taskKey);
+        if (
+          !['chunkSize'].includes(wavKey) &&
+          Object.prototype.hasOwnProperty.call(wav[chunkId], wavKey)
+        ) {
+          wav[chunkId][wavKey] = taskChunk[taskKey];
+        }
+      });
+      resultChunks.push(wav[chunkId]);
+    }
+  });
+
+  console.log(
+    JSON.stringify({
+      msg: 'Wavefile chunks set',
+      chunks: resultChunks,
+    }),
+  );
+
+  // save to s3 destination
+  if (event.Task.Destination.Mode === 'AWS/S3') {
+    await s3Upload(s3, sts, event, Buffer.from(wav.toBuffer()));
+  }
+
+  const now = new Date();
+
+  return {
+    Task: 'WavWrap',
+    Mode: event.Task.Destination.Mode,
+    BucketName: event.Task.Destination.BucketName,
+    ObjectKey: event.Task.Destination.ObjectKey,
+    Time: now.toISOString(),
+    Timestamp: +now / 1000,
+    WavefileChunks: resultChunks,
```

Comment:
Are the output chunks ever different than the values provided in the job? Like is it doing any sort of inspection of the file and provided previously-unknown data? If so, I wonder if it makes more sense to add this to the `Inspect` task, rather than have the WAV wrap task do double duty.

Reply:
It's only showing what was set in the chunks that have been set, which could be different from what is in the job (not for cart chunks, but others), as many of the chunks do have values set based on analyzing the audio.

It is doing an inspection of the file, I wrote an mpeg parser to be able to have the `wavefile` gem correctly set `bext`, `mext', and `fmt` chunks from an mpeg audio file without other dependencies or requiring external knowledge of the file to be passed in, but that data will only come back in this result if any of the values in those chunks are overridden by the task, in which case it will return the combo of the values detected and the overrides applied.

Reply:
also, the inspect task doesn't have this level of info about the details of specific wave chunks, which are in the scope of this task, so I don't think that would be a good fit.